### PR TITLE
Aggregator

### PR DESCRIPTION
**Pull request by boyska**
The aggregator is called `agherant`. It is implemented as a [Blueprint](http://flask.pocoo.org/docs/0.10/blueprints/) for maximum flexibility (it can be reused even without the whole libreant).

The aggregator itself can be aggregated, using `/agherant/search?format=opensearch`.

This is still not optimized:
* no parallel running of query
* no timeouts
* no caching of descriptions

I believe that this can already be merged, and optimizations be postponed.

**Testing**

If you want to test this branch, you will always get no results from agherant. This is because you have to specify which nodes you want to aggregate. You can do it like this:
```sh
WEBANT_AGHERANT_DESCRIPTIONS='["http://127.0.0.1:5000/description.xml"]' webant
```
and you will be aggregating your own node.

Please note that you have to cope with shell quoting, and that the variable is interpreted as json.

### CODE REVIEW

**Comment by boyska** [comment]
**IP/Port**: Yes, is gevent that caused this. Is that a problem? If that's needed, a simple print sounds easy to do ;)

**README**: thanks, fixed

**URL and localhost**: I think this is a non-problem. Using 127.0.0.1 in a public website is bogus, it only makes sense for development. If your website is `my.library.org`, just use `http://my.library.org/description.xml` instead of `http://127.0.0.1/description.xml` and everything will work fine.

**folder for blueprint**: I believe you're right. Do I need to do it now, or can we defer it a bit? I'm quite busy in these days, but it would be a pity to postpone release date.

**misc**: please note that I cleaned up a bit configuration handling. This especially means that lists follow the flask convention of being json-formatted, not `:`-splitted. I'm thinking of `PRESET_PATHS` (see #7 )

**Comment by ael-code** [comment]
> IP/Port: Yes, is gevent that caused this. Is that a problem? If that's needed, a simple print sounds easy to do ;)

not a real problem
> README: thanks, fixed

oke

> URL and localhost: I think this is a non-problem. Using 127.0.0.1 in a public website is bogus, it only makes sense for development. If your website is my.library.org, just use http://my.library.org/description.xml instead of http://127.0.0.1/description.xml and everything will work fine.

I think this is a real problem :). If you can reach your agherant node in multiple way ( lan, pulbic ip, pulblic hostname etc..) it is very limiting that agherant could return results only with one static presetted address. In addition if you change ip or hostname to the machine that runs the node, you have to remember to change the agherant config!! WTF.
There is a simple solution:
when agherant cycles on the addresses to aggregate, if it find a (localhost or 127.0.0.1) it will substitute them with the hostname/ip with which the client have done the request.

> folder for blueprint: I believe you're right. Do I need to do it now, or can we defer it a bit? I'm quite busy in these days, but it would be a pity to postpone release date.

we could open an issue and do it later

> misc: please note that I cleaned up a bit configuration handling. This especially means that lists follow the flask convention of being json-formatted, not :-splitted. I'm thinking of PRESET_PATHS (see #7 )

there should be no problems. PresetManager accepts a list. I like the new configuration handling.

Sorry for my awful English.

**Comment by boyska** [comment]
ok, so the only thing to solve is the hostname part. I think that noone expects that if a box changes hostname than everything will work smoothly. It's never like that.
But ok, let's say we want to do that. What should we replace "localhost" with? Our FQDN? The ip address we are receiving the query on? The ip solution is wrong in case of NAT (which is very common). FQDN sounds fine... if it does resolve on the network dns.

Unless you come up with a more intelligent solution, I think that any application-level solution to the problem will be half-broken, and that the proper solution is always using names. Names are _the_ way of decoupling pointers from ip address, and the coupling can be different in Internet, local network, localhost.

**Comment by ael-code** [comment]
We could use the ```Host``` field of the http request received and if it's not present we will fallback to the ip one.
We are going to interfere in the aggregation process only if we will found ( localhost or 127.0.0.1 ), but we are not taking away the ability to the admin to use a FQDN in AGHERANT_DESCRIPTIONS.  
I agree with you:
>application-level solution to the problem will be half-broken

but this solution is very simple, and without it, the AGHERANT_DESCRIPTIONS will not work against localhost aggregation. Better half-broken then fully-broken ;)

File: webant/webant.py [hunk 1]
```diff
@@ -21,19 +22,17 @@ def initLoggers():
...
-    AppConfig(app, configfile)
+    app.config.update({
+        'BOOTSTRAP_SERVE_LOCAL': True,
+        'DEBUG': True,
+        'PRESET_PATHS': [],
+        'AGHERANT_DESCRIPTIONS': [],
+        'SECRET_KEY': 'really insecure, please change me!'
+    })
+    AppConfig(app, configfile, default_settings=False)
+    app.register_blueprint(agherant, url_prefix='/agherant')
```
Comment (ael-code) [inline webant/webant.py]: Could we register agherant on the flask app only if ```AGHERANT_DESCRIPTIONS``` is not empty?

**Comment by boyska** [comment]
Ok, the `Host` field seems a nice "solution".

Instead of looking for `localhost` what do you think about providing a special `SELF` description that will auto-expand to `url_for('description')` ? This way the auto-aggregate thing is much simpler, and SELF will be the only request where we need to fiddle with the `Host` header. I think this is more explicit, which is better than implicit ;)

File: webant/webant.py (hunk 1 above)
Comment (boyska) [inline webant/webant.py]: yes, that sounds reasonable

**Comment by ael-code** [comment]
nice one. I like it.

I'm curious on how you will get the ip of the node from Flask.

**Comment by boyska** [comment]
I think that I solved all the points that you raised (and something more):
1. the `SELF` things automatically provides the `req.http.Host`, as it uses `url_for`, with `_external=True`, which uses req.http.Host for the host name
2. agherant only loaded if `AGHERANT_DESCRIPTIONS` is not empty

Plus, it adds these nice bounties:
1. there is a `agherant` executable, so you can run only agherant now. This can be useful to run a node which only is an aggregator without a libreant, or just for testing purposes.
2. before, at every request every description url was fetched, and a `Client` object was created. Now this is cached, which will decrease latency. The cache is not the best one you can imagine, since it never expires. Anyway I think it is a good improvement.

Am I missing something?

**Comment by ael-code** [comment]
merged with rebase

**Pull request closed without merging**